File: tests/Aiko.Core.Tests/SettingsJsonPatchTests.cs
using System.Text.Json.Nodes;

namespace Aiko.Core.Tests;

public class SettingsJsonPatchTests
{
    private const string Bridge = @"""C:\Users\slayu\AppData\Local\Aiko\current\Aiko.Bridge.exe""";

    private const string PlainSettings = """
        {
          "$schema": "https://json.schemastore.org/claude-code-settings.json",
          "model": "opusplan",
          "theme": "dark"
        }
        """;

    private const string SettingsWithOwnStatusLine = """
        {
          "model": "opusplan",
          "statusLine": { "type": "command", "command": "~/.claude/my-status.sh" }
        }
        """;

    [Fact]
    public void Adds_our_status_line_and_keeps_everything_else()
    {
        Assert.True(SettingsJsonPatch.TryAddBridge(PlainSettings, Bridge, out var patched));

        var root = JsonNode.Parse(patched)!.AsObject();
        Assert.Equal("opusplan", root["model"]!.GetValue<string>());
        Assert.Equal("dark", root["theme"]!.GetValue<string>());
        Assert.Equal(Bridge, root["statusLine"]!["command"]!.GetValue<string>());
        Assert.Equal("command", root["statusLine"]!["type"]!.GetValue<string>());
    }

    [Fact]
    public void Keeps_the_status_line_the_user_already_had()
    {
        SettingsJsonPatch.TryAddBridge(SettingsWithOwnStatusLine, Bridge, out var patched);

        var root = JsonNode.Parse(patched)!.AsObject();
        Assert.Equal(Bridge, root["statusLine"]!["command"]!.GetValue<string>());
        Assert.Equal("~/.claude/my-status.sh", root[SettingsJsonPatch.WrappedKey]!["command"]!.GetValue<string>());
    }

    [Fact]
    public void The_bridge_can_read_the_command_it_has_to_call()
    {
        SettingsJsonPatch.TryAddBridge(SettingsWithOwnStatusLine, Bridge, out var patched);

        Assert.Equal("~/.claude/my-status.sh", SettingsJsonPatch.ReadWrappedCommand(patched));
        Assert.Null(SettingsJsonPatch.ReadWrappedCommand(PlainSettings));
    }

    [Fact]
    public void Adding_twice_changes_nothing()
    {
        SettingsJsonPatch.TryAddBridge(PlainSettings, Bridge, out var once);

        Assert.False(SettingsJsonPatch.TryAddBridge(once, Bridge, out var twice));
        Assert.Equal(once, twice);
    }

    [Fact]
    public void Removing_puts_the_file_back_the_way_it_was()
    {
        SettingsJsonPatch.TryAddBridge(PlainSettings, Bridge, out var patched);

        Assert.True(SettingsJsonPatch.TryRemoveBridge(patched, out var restored));

        var root = JsonNode.Parse(restored)!.AsObject();
        Assert.False(root.ContainsKey("statusLine"));
        Assert.False(root.ContainsKey(SettingsJsonPatch.WrappedKey));
        Assert.Equal("opusplan", root["model"]!.GetValue<string>());
    }

    [Fact]
    public void Removing_gives_the_user_their_own_status_line_back()
    {
        SettingsJsonPatch.TryAddBridge(SettingsWithOwnStatusLine, Bridge, out var patched);

        SettingsJsonPatch.TryRemoveBridge(patched, out var restored);

        var root = JsonNode.Parse(restored)!.AsObject();
        Assert.Equal("~/.claude/my-status.sh", root["statusLine"]!["command"]!.GetValue<string>());
        Assert.False(root.ContainsKey(SettingsJsonPatch.WrappedKey));
    }

    [Fact]
    public void Removing_from_a_file_we_never_touched_changes_nothing()
    {
        Assert.False(SettingsJsonPatch.TryRemoveBridge(PlainSettings, out var restored));
        Assert.Equal(PlainSettings, restored);
    }

    [Theory]
    [InlineData("")]
    [InlineData("not json")]
    [InlineData("[1, 2]")]
    public void A_file_we_cannot_read_is_left_alone(string json)
    {
        Assert.False(SettingsJsonPatch.TryAddBridge(json, Bridge, out var patched));
        Assert.Equal(json, patched);
        Assert.False(SettingsJsonPatch.TryRemoveBridge(json, out var restored));
        Assert.Equal(json, restored);
    }

    [Fact]
    public void An_empty_command_is_refused()
    {
        Assert.False(SettingsJsonPatch.TryAddBridge(PlainSettings, "   ", out var patched));
        Assert.Equal(PlainSettings, patched);
    }
}

File: mac/Tests/AikoKitTests/SettingsJsonPatchTests.swift
import Foundation
import Testing

@testable import AikoKit

struct SettingsJsonPatchTests {
    static let bridge = #""C:\Users\slayu\AppData\Local\Aiko\current\Aiko.Bridge.exe""#

    static let plainSettings = """
        {
          "$schema": "https://json.schemastore.org/claude-code-settings.json",
          "model": "opusplan",
          "theme": "dark"
        }
        """

    static let settingsWithOwnStatusLine = """
        {
          "model": "opusplan",
          "statusLine": { "type": "command", "command": "~/.claude/my-status.sh" }
        }
        """

    private var patch: SettingsJsonPatch { SettingsJsonPatch(.windows) }

    @Test
    func addsOurStatusLineAndKeepsEverythingElse() throws {
        let patched = try #require(patch.addBridge(Self.plainSettings, Self.bridge))

        let root = try #require(JsonNode.parse(patched)?.objectValue)
        #expect(root["model"]?.stringValue == "opusplan")
        #expect(root["theme"]?.stringValue == "dark")
        #expect(root["statusLine"]?["command"]?.stringValue == Self.bridge)
        #expect(root["statusLine"]?["type"]?.stringValue == "command")
    }

    @Test
    func keepsTheStatusLineTheUserAlreadyHad() throws {
        let patched = try #require(patch.addBridge(Self.settingsWithOwnStatusLine, Self.bridge))

        let root = try #require(JsonNode.parse(patched)?.objectValue)
        #expect(root["statusLine"]?["command"]?.stringValue == Self.bridge)
        #expect(root[SettingsJsonPatch.wrappedKey]?["command"]?.stringValue == "~/.claude/my-status.sh")
    }

    @Test
    func theBridgeCanReadTheCommandItHasToCall() throws {
        let patched = try #require(patch.addBridge(Self.settingsWithOwnStatusLine, Self.bridge))

        #expect(SettingsJsonPatch.readWrappedCommand(patched) == "~/.claude/my-status.sh")
        #expect(SettingsJsonPatch.readWrappedCommand(Self.plainSettings) == nil)
    }

    @Test
    func addingTwiceChangesNothing() throws {
        let once = try #require(patch.addBridge(Self.plainSettings, Self.bridge))

        #expect(patch.addBridge(once, Self.bridge) == nil)
    }

    @Test
    func removingPutsTheFileBackTheWayItWas() throws {
        let patched = try #require(patch.addBridge(Self.plainSettings, Self.bridge))

        let restored = try #require(patch.removeBridge(patched))

        let root = try #require(JsonNode.parse(restored)?.objectValue)
        #expect(!root.contains("statusLine"))
        #expect(!root.contains(SettingsJsonPatch.wrappedKey))
        #expect(root["model"]?.stringValue == "opusplan")
    }

    @Test
    func removingGivesTheUserTheirOwnStatusLineBack() throws {
        let patched = try #require(patch.addBridge(Self.settingsWithOwnStatusLine, Self.bridge))

        let restored = try #require(patch.removeBridge(patched))

        let root = try #require(JsonNode.parse(restored)?.objectValue)
        #expect(root["statusLine"]?["command"]?.stringValue == "~/.claude/my-status.sh")
        #expect(!root.contains(SettingsJsonPatch.wrappedKey))
    }

    @Test
    func removingFromAFileWeNeverTouchedChangesNothing() {
        #expect(patch.removeBridge(Self.plainSettings) == nil)
    }

    @Test(arguments: ["", "not json", "[1, 2]"])
    func aFileWeCannotReadIsLeftAlone(json: String) {
        #expect(patch.addBridge(json, Self.bridge) == nil)
        #expect(patch.removeBridge(json) == nil)
    }

    @Test
    func anEmptyCommandIsRefused() {
        #expect(patch.addBridge(Self.plainSettings, "   ") == nil)
    }

    @Test(arguments: [
        (#"{ "outputStyle": "Explanatory" }"#, "Explanatory"),
        (#"{ "outputStyle": "my-plugin:Terse" }"#, "my-plugin:Terse"),
        (#"{ "outputStyle": "default" }"#, nil),
        (#"{ "outputStyle": "Aiko" }"#, nil),
        (#"{ "outputStyle": "aiko-persona:Aiko" }"#, nil),
        (#"{ "outputStyle": "" }"#, nil),
        (#"{ "outputStyle": 3 }"#, nil),
        (#"{ "model": "opus" }"#, nil),
        ("not json", nil),
    ] as [(String, String?)])
    func onlyAStyleThePersonPickedCountsAsTheirOwn(json: String, expected: String?) {
        #expect(SettingsJsonPatch.userOutputStyle(json) == expected)
    }

    // ---- the session start hook, from SessionReminderTests ----

    static let hook = #""C:\Program Files\Aiko\Aiko.Bridge.exe" --session-start"#

    @Test
    func theHookIsAddedBesideThePersonsOwnHooks() throws {
        let before = """
            {
              "model": "opus",
              "hooks": {
                "SessionStart": [ { "matcher": "startup", "hooks": [ { "type": "command", "command": "echo hi" } ] } ],
                "Stop": [ { "hooks": [ { "type": "command", "command": "notify" } ] } ]
              }
            }
            """

        let after = try #require(patch.addSessionHook(before, Self.hook))

        let groups = try #require(JsonNode.parse(after)?["hooks"]?["SessionStart"]?.arrayValue)
        #expect(groups.count == 2)
        #expect(groups[0]["hooks"]?.arrayValue?[0]["command"]?.stringValue == "echo hi")
        #expect(groups[1]["hooks"]?.arrayValue?[0]["command"]?.stringValue == Self.hook)
        #expect(JsonNode.parse(after)?["hooks"]?["Stop"] != nil)
        #expect(patch.hasOurSessionHook(after))
    }

    @Test
    func addingTwiceChangesNothingAndAnOldPathIsReplaced() throws {
        let once = try #require(patch.addSessionHook("{}", Self.hook))
        #expect(patch.addSessionHook(once, Self.hook) == nil)

        let moved = #""D:\Aiko\Aiko.Bridge.exe" --session-start"#
        let replaced = try #require(patch.addSessionHook(once, moved))

        let commands = try #require(JsonNode.parse(replaced)?["hooks"]?["SessionStart"]?.arrayValue)
        #expect(commands.count == 1)
        #expect(commands[0]["hooks"]?.arrayValue?[0]["command"]?.stringValue == moved)
    }

    @Test
    func removingAikoTakesTheHookOutAndLeavesTheFileAsItWas() throws {
        let before = #"{ "hooks": { "SessionStart": [ { "hooks": [ { "type": "command", "command": "echo hi" } ] } ] } }"#

        let withLine = try #require(patch.addBridge(before, #""C:\A\Aiko.Bridge.exe""#))
        let withBoth = try #require(patch.addSessionHook(withLine, Self.hook))

        let restored = try #require(patch.removeBridge(withBoth))
        #expect(JsonNode.parse(restored) == JsonNode.parse(before))
    }

    @Test
    func aFileWithOnlyOurHookLosesTheEmptyHooksObject() throws {
        let withHook = try #require(patch.addSessionHook(#"{ "model": "opus" }"#, Self.hook))

        let restored = try #require(patch.removeBridge(withHook))
        #expect(JsonNode.parse(restored) == JsonNode.parse(#"{ "model": "opus" }"#))
    }

    @Test
    func hooksThatAreNotTheExpectedShapeAreNotTouched() {
        #expect(patch.addSessionHook(#"{ "hooks": "oops" }"#, Self.hook) == nil)
        #expect(patch.addSessionHook(#"{ "hooks": { "SessionStart": {} } }"#, Self.hook) == nil)
    }

    @Test
    func theHookCommandIsTheBridgeWithOneArgument() {
        #expect(BridgeCommand.hookFor(#"C:\A\Aiko.Bridge.exe"#) == #""C:\A\Aiko.Bridge.exe" --session-start"#)
        #expect(
            BridgeCommand.hookFor(#"C:\A\Aiko.Bridge.exe"#, .powerShell)
                == #"& "C:\A\Aiko.Bridge.exe" --session-start"#)
        #expect(BridgeCommand.isAiko(.windows, BridgeCommand.hookFor(#"C:\A\Aiko.Bridge.exe"#)))
    }
}
